src/core/qtlibtorrent/qbtsession.cpp

```diff
@@ -1276,72 +1276,78 @@ void QBtSession::loadTorrentTempData(QTorrentHandle &h, QString savePath, bool m
-    QList<QUrl> new_trackers = misc::magnetUriToTrackers(magnet_uri);
-    bool trackers_added = false;
-    foreach (const QUrl& new_tracker, new_trackers) {
+    QStringList trackers = misc::magnetUriToTrackers(magnet_uri);
+    QStringList urlSeeds = misc::magnetUriToUrlSeeds(magnet_uri);
+    mergeTorrents_impl(h, trackers, urlSeeds);
+}
+
+void QBtSession::mergeTorrents(const QTorrentHandle &h, const boost::intrusive_ptr<torrent_info> t) {
+    QStringList trackers;
+    QStringList urlSeeds;
+
+    foreach (const announce_entry& newTracker, t->trackers())
+        trackers.append(misc::toQStringU(newTracker.url));
+
+    foreach (const web_seed_entry& newUrlSeed, t->web_seeds())
+        urlSeeds.append(misc::toQStringU(newUrlSeed.url));
+
+    mergeTorrents_impl(h, trackers, urlSeeds);
+}
+
+void QBtSession::mergeTorrents_impl(const QTorrentHandle &h, const QStringList &trackers, const QStringList &urlSeeds)
+{
+    if (!h.is_valid())
+        return;
+
+    QString hash = h.hash();
+    QString name = h.name();
+    const std::vector<announce_entry> existingTrackers = h.trackers();
+    const QStringList existingUrlSeeds = h.url_seeds();
+
+    foreach (const QString &tracker, trackers) {
+        QUrl trackerUrl(tracker);
         bool found = false;
-        std::vector<announce_entry> existing_trackers = h_ex.trackers();
-        foreach (const announce_entry& existing_tracker, existing_trackers) {
-            if (new_tracker == QUrl(existing_tracker.url.c_str())) {
+
+        foreach (const announce_entry &existingTracker, existingTrackers) {
+            QUrl existingTrackerUrl(misc::toQStringU(existingTracker.url));
+            if (trackerUrl == existingTrackerUrl) {
                 found = true;
                 break;
             }
         }
 
         if (!found) {
-            h_ex.add_tracker(announce_entry(new_tracker.toString().toStdString()));
-            trackers_added = true;
+            h.add_tracker(announce_entry(tracker.toUtf8().constData()));
+            emit trackerAdded(tracker, hash);
+            Logger::instance()->addMessage(tr("Tracker '%1' was added to torrent '%2'").arg(tracker).arg(name));
         }
     }
-    if (trackers_added)
-        Logger::instance()->addMessage(tr("Note: new trackers were added to the existing torrent."));
-}
 
-void QBtSession::mergeTorrents(QTorrentHandle &h_ex, boost::intrusive_ptr<torrent_info> t) {
-    // Check if the torrent contains trackers or url seeds we don't know about
-    // and add them
-    if (!h_ex.is_valid()) return;
-    std::vector<announce_entry> existing_trackers = h_ex.trackers();
-    std::vector<announce_entry> new_trackers = t->trackers();
-    bool trackers_added = false;
-    foreach (const announce_entry& new_tracker, new_trackers) {
-        std::string new_tracker_url = new_tracker.url;
-        // Check if existing torrent has this tracker
+    if (existingTrackers.empty() && !h.trackers().empty())
+        emit trackerRemoved("", hash);
```

Comment:
> There is a trackerlessStateChanged(bool trackerless) signal, which is sent when this state is changed.

So basically perform the same checks I already do in the diff. But instead of emiting an "obscure" signal `trackerRemoved("", hash)` I should emit a more descriptive one?